src/renderers/dom/shared/DOMProperty.js

```diff
@@ -197,33 +198,83 @@ var DOMProperty = {
   properties: {},
 
   /**
-   * Mapping from lowercase property names to the properly cased version, used
-   * to warn in the case of missing properties. Available only in __DEV__.
-   *
-   * autofocus is predefined, because adding it to the property whitelist
-   * causes unintended side effects.
-   *
-   * @type {Object}
+   * Some attributes are aliased for easier use within React. We don't
+   * allow direct use of these attributes. See DOMAttributeNames in
+   * HTMLPropertyConfig and SVGPropertyConfig.
    */
-  getPossibleStandardName: __DEV__ ? {autofocus: 'autoFocus'} : null,
+  aliases: {},
 
   /**
-   * All of the isCustomAttribute() functions that have been injected.
-   */
-  _isCustomAttributeFunctions: [],
-
-  /**
-   * Checks whether a property name is a custom attribute.
+   * Checks whether a property name is a writeable attribute.
    * @method
    */
-  isCustomAttribute: function(attributeName) {
-    for (var i = 0; i < DOMProperty._isCustomAttributeFunctions.length; i++) {
-      var isCustomAttributeFn = DOMProperty._isCustomAttributeFunctions[i];
-      if (isCustomAttributeFn(attributeName)) {
+  shouldSetAttribute: function(name, value) {
+    if (DOMProperty.isReservedProp(name)) {
+      return false;
+    }
+
+    if (value === null) {
+      return true;
+    }
+
+    var lowerCased = name.toLowerCase();
+
+    // Prevent aliases, and badly cased aliases like `class` or `cLASS`
+    // from showing up in the DOM
+    if (DOMProperty.aliases.hasOwnProperty(lowerCased)) {
+      return false;
+    }
+
+    var propertyInfo = DOMProperty.properties[name];
+
+    switch (typeof value) {
+      case 'boolean':
+        if (propertyInfo) {
+          return true;
+        }
+        var prefix = lowerCased.slice(0, 5);
+        return prefix === 'data-' || prefix === 'aria-';
+      case 'undefined':
+      case 'number':
+      case 'string':
         return true;
-      }
+      case 'object':
+        // Allow HAS_BOOLEAN_VALUE to coerce to true
+        if (propertyInfo && propertyInfo.hasBooleanValue) {
+          return true;
+        }
+        if (name === 'value') {
+          // Value is a special prop.
+          // Allow arrays or objects with custom toString.
+          return value.toString !== Object.prototype.toString;
+        }
+        // For everything else, allow only objects with custom toString.
+        if (Array.isArray(value)) {
+          return false;
+        }
+        return value.toString !== Object.prototype.toString;
+      default:
+        return false;
     }
-    return false;
+  },
```

Comment:
@sebmarkbage Follow up from https://github.com/facebook/react/pull/10416 regarding valueOf.

Is it safe to evaluate stringification through `valueOf` or `toString`, and fallback to `null` if we don't like the value?

Like:

```javascript
var stringified = '' + value
return stringified !== '[object Object]'
```

I imagine just checking for `[object Object]` isn't the way to go, but something similar.

We could change this function to be more of a "getAttributeValue" function, which would return `null` in the case where an attribute should be assigned. That way if a `valueOf` or `toString`, _heaven forbid_, have side-effects, we only call it once.